File: job-addon-api/src/main/java/com/frost_fox/jenkins/job_addon/Result.java
package com.frost_fox.jenkins.job_addon;

public class Result<T> {

    private final boolean success;
    private final T item;
    private final String cause;

    private Result(boolean success, T item, String cause) {
        this.success = success;
        this.item = item;
        this.cause = cause;
    }

    @SuppressWarnings("unused")
    public boolean isSuccess() {
        return success;
    }

    @SuppressWarnings("unused")
    public T getItem() {
        return item;
    }


    public String getCause() {
        return cause;
    }

    public static Result failedWith(String cause) {
        return new Result<>(false, null, cause);
    }

    public static <E> Result<E> successfulWith(E item) {
        return new Result<>(true, item, "");
    }

    public boolean successful() {
        return success;
    }

    public T get() {
        return item;
    }
}

File: job-addon-grid/src/main/frontend/src/job/integration/Result.js
export default class Result {

    /**
     * @param {boolean} success
     * @param {any} item
     * @param {string} cause
     */
    constructor(success, item, cause) {
        this.success = success;
        this.item = item;
        this.cause = cause;
    }

    static successful(item) {
        return new Result(true, item, "");
    }

    static failed(cause) {
        return new Result(false, null, cause);
    }

    static fromResponse(response) {
        this.validateStructure(response);
        return new this(response.success, response.item, response.cause);
    }

    static validateStructure(response) {
        if (typeof response.success !== "boolean" || response.item === undefined || typeof response.cause !== "string") {
            throw new Error("unexpected response structure");
        }
    }

    isSuccessful() {
        return this.success;
    }

    get() {
        return this.item;
    }

    getCause() {
        return this.cause;
    }
}